in_toto/artifact_rules.py

```diff
@@ -0,0 +1,186 @@
+        " rule must be a list.\n"
+        "Got: \n\t'{0}'".format(rule))
+
+  rule_len = len(rule)
+
+  if (rule_len < 2 or not isinstance(rule[0], six.string_types) or
+      rule[0].lower() not in ALL_RULES):
+    raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+        " rule must start with one of '{0}' and specify a 'pattern' as"
+        " second element.\n"
+        "Got: \n\t'{1}'".format(", ".join(ALL_RULES), rule))
+
+  rule_type = rule[0].lower()
+  pattern = rule[1]
+
+  if not isinstance(pattern, six.string_types):
+    raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+        " 'pattern' (second element) must be a string.\n"
+        "Got: \n\t'{0}'".format(rule))
+
+  # Type is one of "CREATE", "MODIFY", "DELETE", "ALLOW", "DISALLOW"
+  if rule_type in GENERIC_RULES:
+    if rule_len != 2:
+      raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+        " generic rules must have one of the formats:\n\t"
+        "CREATE <pattern>\n\t"
+        "MODIFY <pattern>\n\t"
+        "DELETE <pattern>\n\t"
+        "ALLOW <pattern>\n\t"
+        "DISALLOW <pattern>\n"
+        "Got:\n\t{}".format(rule))
+    else:
+      return {
+        "type": rule_type,
+        "pattern": pattern,
+      }
+
+  # Type is "MATCH"
+  elif rule_type in COMPLEX_RULES:
+
+    # ... IN <source-path-prefix> WITH (MATERIALS|PRODUCTS)
+    # IN <destination-path-prefix> FROM <step>
+    if (rule_len == 10 and rule[2].lower() == "in" and
+        rule[4].lower() == "with" and rule[6].lower() == "in" and
+        rule[8].lower() == "from"):
+      source_prefix = rule[3]
+      dest_type = rule[5]
+      dest_prefix = rule[7]
+      dest_name = rule[9]
+
+    # ... IN <source-path-prefix> WITH (MATERIALS|PRODUCTS) FROM <step>
+    elif (rule_len == 8 and rule[2].lower() == "in" and
+        rule[4].lower() == "with" and rule[6].lower() == "from"):
+      source_prefix = rule[3]
+      dest_type = rule[5]
+      dest_prefix = ""
+      dest_name = rule[7]
+
+    # ... WITH (MATERIALS|PRODUCTS) IN <destination-path-prefix> FROM <step>
+    elif (rule_len == 8 and rule[2].lower() == "with" and
+        rule[4].lower() == "in" and rule[6].lower() == "from"):
+      source_prefix = ""
+      dest_type = rule[3]
+      dest_prefix = rule[5]
+      dest_name = rule[7]
+
+    # ... WITH (MATERIALS|PRODUCTS) FROM <step>
+    elif (rule_len == 6 and rule[2].lower() == "with" and
+        rule[4].lower() == "from"):
+      source_prefix = ""
+      dest_type = rule[3]
+      dest_prefix = ""
+      dest_name = rule[5]
+
+    else:
+      raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+          " match rules must have the format:\n\t"
+          " MATCH <pattern> [IN <source-path-prefix>] WITH"
+          " (MATERIALS|PRODUCTS) [IN <destination-path-prefix>] FROM <step>.\n"
+          "Got: \n\t{}".format(rule))
+
+    if not isinstance(dest_type, six.string_types) or (
```

Comment:
wouldn't it be nicer to iterate over all the rules to check they are strings first?